bpf/lib/ipv4.h

```diff
@@ -44,8 +68,71 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
 	// +----+----+----+----------------------------------+
 	//
 	// If "More fragments" or the offset is nonzero, then this is an IP
-	// fragment. The evil bit must be set to 0 (RFC791, RFC3514).
-	return ip4->frag_off & bpf_htons(0xBFFF);
+	// fragment (RFC791).
+	return ip4->frag_off & bpf_htons(0x3FFF);
+}
+
+#if defined IPV4_FRAGMENTS
+static __always_inline bool ipv4_is_not_first_fragment(const struct iphdr *ip4)
+{
+	/* Ignore "More fragments" bit to catch all fragments but the first */
+	return ip4->frag_off & bpf_htons(0x1FFF);
+}
+
+static __always_inline int
+ipv4_frag_get_l4ports(const struct ipv4_frag_id *frag_id,
+		      struct ipv4_frag_l4ports *ports)
+{
+	struct ipv4_frag_l4ports *tmp;
+
+	tmp = map_lookup_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id);
+	if (!tmp)
+		return DROP_FRAG_NOT_FOUND;
+
+	/* Do not make ports a pointer to map data, copy from map */
+	__builtin_memcpy(ports, tmp, sizeof(*ports));
+	return 0;
+}
+
+static __always_inline int
+ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off,
+			    const struct ipv4_frag_id *frag_id,
+			    struct ipv4_frag_l4ports *ports)
+{
+	int ret;
+
+	ret = ctx_load_bytes(ctx, l4_off, ports, 4);
+	if (ret < 0)
+		return ret;
+
+	map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY);
```

Comment:
Can we add a metric to the Cilium metrics map to have some introspection that fragments were seen? Would be super useful to have this visible in Prometheus as a graph later to see when/if potential configuration issues appeared.

Reply:
Maybe also a metric if the map update failed to be able to debug that case when it could not be handled.